File: Training/UnitTesting/Python/Test Driven Development/Exercises/after/bowling/game.py
class Game:
    def __init__(self):
        self.rolls = []
    
    def roll(self, pins):
        self.rolls.append(pins)
    
    def score(self):
        score = 0
        i = 0
        for _ in range(10):
            next_two = self.rolls[i] + self.rolls[i+1]
            score += next_two
            if (next_two >= 10):
                score += self.rolls[i+2]
            i += 1 if self.rolls[i] == 10 else 2
        return score

File: Training/UnitTesting/C#/Test Driven Development/Exercises/after/Bowling/Game.cs
using System;

namespace Bowling
{
    public class Game
    {
        private int[] rolls = new int[21];
        private int rollIndex = 0;

        public int Score() 
        {
            int score = 0;
            int frameIndex = 0;
            for (int frame = 0; frame < 10; frame++)
            {
                score += NormalScore(frameIndex);
                if (IsStrike(frameIndex) || IsSpare(frameIndex))
                {
                    score += Bonus(frameIndex);
                }
                frameIndex += (IsStrike(frameIndex) ? 1 : 2);
            }
            return score;
        }

        private bool IsSpare(int i) 
        {
            return rolls[i] + rolls[i+1] == 10;
        }

        private bool IsStrike(int i)
        {
            return rolls[i] == 10;
        }

        private int NormalScore(int i) 
        {
            return rolls[i] + rolls[i+1];
        }

        private int Bonus(int i)
        {
            return rolls[i+2];
        }

        public void Roll(int pins)
        {
            rolls[rollIndex] = pins;
            rollIndex++;
        }
    }
}
